drivers: stepper: step_dir: fix 87698

jilaypandya opened the pull request:
During testing with teensy 4.0 it was discoverd that toggling at high frequencies led to missed steps. 
As per the datasheets of step-dir drivers, an active edge needs to be maintained for a certain time period, for a step to be executed.

jilaypandya committed:
drivers: stepper: step_dir: fix 87698

During testing with teensy 4.0 it was discoverd that toggling
at high frequencies led to missed steps. As per the datasheets
of step-dir drivers an active edge needs to be maintained for
a certain timerperiod, for a step to be executed.

Instead of toggling pin twice instantaneously, the current fix
halves the step period and the step-pin is toggled on the timeout
of the respective timing source.

rework stepper stepper stop

set step pin to low if the driver does not support dual edge
and increment/decrement the actual position by one

Signed-off-by: Jilay Pandya <jilay.pandya@outlook.com>

faxe1008 commented on drivers/stepper/step_dir/step_dir_stepper_common.c:
```diff
@@ -8,9 +8,19 @@
+		CONTAINER_OF(timer, struct step_dir_stepper_common_data, step_pulse_timer);
+	const struct step_dir_stepper_common_config *config = data->dev->config;
+
+	gpio_pin_set_dt(&config->step_pin, 0);
```
I'd use 'gpio_pin_toggle_dt` instead here.

faxe1008 commented on dts/bindings/stepper/step-dir-timing.yaml:
```diff
@@ -0,0 +1,12 @@
+    required: true
+    description: |
+      Total pulse width in nanoseconds for the step signal.
+      This is the total duration for which the step signal is asserted high and low.
```
Not sure if the description is correct, this is what the timer effectively does:
<img width="810" height="237" alt="image" src="https://github.com/user-attachments/assets/5ba2fc3f-fcae-4f7d-a4b9-c557d8760800" />

jilaypandya commented on drivers/stepper/step_dir/step_dir_stepper_common.c:
```diff
@@ -8,9 +8,19 @@
+		CONTAINER_OF(timer, struct step_dir_stepper_common_data, step_pulse_timer);
+	const struct step_dir_stepper_common_config *config = data->dev->config;
+
+	gpio_pin_set_dt(&config->step_pin, 0);
```
with the counter approach, the `stepper_handle_timing_signal` is also getting called in ISR context. Would adding a task to a dedicated workq from the `stepper_handle_timing_signal` be the correct approach?

jbehrensnx replied:
I did some tests on a Nucleo WB55RG. The `k_timer` does not offer sufficient accuracy, at least when using a counter timing source. As a result, the delay function needs to use the same timing source as the main step function.

A toggling bool to check for pin high or pin low in combination with a if check in `stepper_handle_timing_signal` for setting the pin to low might work for this.

The delays from using a workqueue are as bad as using a `k_timer`. As a result, the gpio functions need to be called in an ISR context, despite the issues from doing this.

mbolivar reviewed (dismissed):
+1 is only for the bindings

dipakgmx commented on drivers/stepper/step_dir/step_dir_stepper_common.h:
```diff
@@ -78,7 +80,7 @@ struct step_dir_stepper_common_data {
 
 	struct k_work_delayable stepper_dwork;
-
+	bool step_completed;
```
This too needs to be an atomic variable

dipakgmx commented on drivers/stepper/step_dir/step_dir_stepper_common.c:
```diff
@@ -335,8 +341,11 @@ int step_dir_stepper_common_run(const struct device *dev, const enum stepper_dir
+	struct step_dir_stepper_common_data *data = dev->data;
 	int ret;
 
+	k_sleep(K_NSEC(config->step_width_ns));
```
If am not wrong, this is the user facing API. Any reason to wait here 20ns?

jilaypandya replied:
I have reworked the stop function and removed the delay.

1. stop timing source.
2. trigger stop event.
3. toggle the step pin to low if required.

dipakgmx commented on drivers/stepper/step_dir/step_dir_stepper_common.h:
```diff
@@ -74,11 +77,12 @@ struct step_dir_stepper_common_data {
 
 	struct k_work_delayable stepper_dwork;
-
+	atomic_t step_completed;
```
Nit: This could be further optimised by using a bit array.

```
enum {
	STEP_DIR_STEPPER_STEP_COMPLETED,
	STEP_DIR_STEPPER_STOPPED,
	...
	STEP_DIR_STEPPER_FLAG_NUM_FLAGS,
};


ATOMIC_DEFINE(flags, STEP_DIR_STEPPER_FLAG_NUM_FLAGS);


// Step complete for example
atomic_set_bit(dev->flags, STEP_DIR_STEPPER_STEP_COMPLETED);
```

jilaypandya committed:
tests: stepper_api: adjust timing timeout tolerance

Adjust CONFIG_STEPPER_TEST_TIMING_TIMEOUT_TOLERANCE_PCT
in order to pass the tests

Signed-off-by: Jilay Pandya <jilay.pandya@outlook.com>

dipakgmx commented on drivers/stepper/step_dir/step_dir_stepper_common.c:
```diff
@@ -159,13 +143,28 @@ static void velocity_mode_task(const struct device *dev)
-	(void)step_dir_stepper_perform_step(dev);
-	if (data->direction == STEPPER_DIRECTION_POSITIVE) {
-		atomic_inc(&data->actual_position);
+	atomic_xor(&data->step_high, 1);
```
Use `atomic_cas` here.

dipakgmx commented on dts/bindings/stepper/allegro/allegro,a4979.yaml:
```diff
@@ -25,8 +25,12 @@ compatible: "allegro,a4979"
 
 properties:
+  step-width-ns:
+    default: 1000
```
```suggestion
    type: int
    default: 1000
```

rruuaanng replied:
Please add a description of the default value.

jbehrensnx commented on tests/drivers/stepper/stepper_api/Kconfig:
```diff
@@ -7,7 +7,7 @@ source "Kconfig.zephyr"
 config STEPPER_TEST_TIMING_TIMEOUT_TOLERANCE_PCT
 	int "Stepper timing tolerance percentage"
-	default 20
+	default 120
```
Revert this. Only the work queue tests require the significantly increased tolerance, and theirs is set in the `testcase.yaml`

jbehrensnx commented on drivers/stepper/step_dir/step_dir_stepper_common.c:
```diff
@@ -335,15 +364,26 @@ int step_dir_stepper_common_run(const struct device *dev, const enum stepper_dir
 		return ret;
 	}
-
 	stepper_trigger_callback(dev, STEPPER_EVENT_STOPPED);
```
There needs to be a delay between stopping the timing source and triggering the callback of `step_width_ns` to ensure that the last step has been properly registered by the stepper. In addition, I would prefer, if the position is updated before the callback is triggered.

A `k_sleep` would be in my opinion sufficient for the required delay.
I am not sure, why you removed this delay, as it was there in a previous iteration (note that I didn't review said iteration, so I cant draw conclusions from context).

jbehrensnx replied:
The use case is for the delay is indeed 2. The goal is to minimize missed steps. If the official statement is that `get_actual_position` doesn't guarantee the correct position in this case, then it does becomes easier and the delay is not necessary.

That said, in that case I would like it mentioned in the documentation in some form that `get_actual_position` can't make guarantees. If I missed that part somewhere, this is of course not necessary.

jilaypandya replied:
Open Loop Systems cannot 100% guarantee the actual state. But yeah you are right, let's put this into documentation, to avoid confusion in the future.

dipakgmx commented on include/zephyr/drivers/stepper.h:
```diff
@@ -354,7 +354,9 @@ static inline int z_impl_stepper_set_reference_position(const struct device *dev
- * @brief Get the actual a.k.a reference position of the stepper
+ * @brief Get the actual step count for a given stepper.
+ * @details This function does not guarantee that the returned position is the exact current
+ * position. For precise positioning, encoders should be used in addition to the stepper driver.
```
Yeah, seem good. Would it be possible to change the `@details` to `@note`. I see the latter often.

jilaypandya replied:
done

rruuaanng commented on dts/bindings/stepper/adi/adi,tmc2209.yaml:
```diff
@@ -19,8 +19,12 @@ compatible: "adi,tmc2209"
 
 properties:
+  step-width-ns:
+    default: 100
```
+1

dipakgmx commented on dts/bindings/stepper/ti/ti,drv84xx.yaml:
```diff
@@ -31,8 +31,12 @@ compatible: "ti,drv84xx"
 
 properties:
+  step-width-ns:
+    default: 970
```
Isn't this already described in the `step-dir-timing.yaml`?
